File: src/main/java/org/dxworks/utils/java/rest/client/PathResolver.java
package org.dxworks.utils.java.rest.client;

import com.google.common.annotations.VisibleForTesting;

import java.util.Arrays;
import java.util.Collections;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class PathResolver {

    private final String apiUrl;

    public PathResolver(String apiUrl) {
        this.apiUrl = apiUrl;
    }

    public String getApiPath(Map<String, String> variableValues, String... pathVariables) {
        return Stream.concat(Stream.of(apiUrl),
                Arrays.stream(pathVariables).map(var -> replacePlaceholders(var, variableValues)))
                .collect(Collectors.joining("/"));
    }

    @VisibleForTesting
    String replacePlaceholders(String var, Map<String, String> variablesValues) {
        if (var.startsWith(":"))
            return variablesValues.getOrDefault(var.replace(":", ""), var);
        return var;
    }

    public String getApiPath(String... pathVariables) {
        return getApiPath(Collections.emptyMap(), pathVariables);
    }
}

File: src/main/kotlin/org/dxworks/utils/java/rest/client/PathResolver.kt
package org.dxworks.utils.java.rest.client

import com.google.common.annotations.VisibleForTesting
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream

open class PathResolver(private val apiUrl: String) {
    fun getApiPath(variableValues: Map<String, String>, vararg pathVariables: String): String {
        return Stream.concat(Stream.of(apiUrl),
            Arrays.stream(pathVariables).map { v: String -> replacePlaceholders(v, variableValues) })
            .collect(Collectors.joining("/"))
    }

    @VisibleForTesting
    fun replacePlaceholders(v: String, variablesValues: Map<String, String>): String {
        return if (v.startsWith(":")) variablesValues.getOrDefault(v.replace(":", ""), v) else v
    }

    fun getApiPath(vararg pathVariables: String): String {
        return getApiPath(emptyMap(), *pathVariables)
    }

}
